core/dbt/deps/tarball.py

```diff
@@ -0,0 +1,185 @@
+import os
+# import shutil
+import hashlib
+import tarfile
+from tempfile import NamedTemporaryFile
+from typing import Optional
+
+from dbt.clients import system
+from dbt.deps.base import PinnedPackage, UnpinnedPackage, get_downloads_path
+from dbt.config import Project
+from dbt.contracts.project import (
+    ProjectPackageMetadata,
+    TarballPackage,
+)
+
+from dbt.logger import GLOBAL_LOGGER as logger
+
+TARFILE_MAX_SIZE = 1 * 1e+6  # limit tarfiles to 1mb
+
+
+class TarballPackageMixin:
+    def __init__(self, tarball: str) -> None:
+        super().__init__()
+        self.tarball = tarball
+
+    @property
+    def name(self):
+        return self.tarball
+
+    def source_type(self) -> str:
+        return 'tarball'
+
+
+class TarballPinnedPackage(TarballPackageMixin, PinnedPackage):
+    def __init__(
+        self, 
+        tarball: str,
+        sha1: Optional[str] = None,
+        subdirectory: Optional[str] = None,
+    ) -> None:
+        super().__init__(tarball)
+        self.sha1 = sha1
+        self.subdirectory = subdirectory
+        self.tarfile = self.get_tarfile()
+        # init tarfile in class, and use tarfile as cache
+        # simpler for tempfile handling, and prevent multiple hits to
+        # url (one for metadata, one for intall)
+        self.tar_dir_name = self.resolve_tar_dir()
+        # assumed structure is that tarfile has a root dir (package name)
+        # but we don't know what it is. will scan file for best candidate
+        # todo implement subdirectory like in git package
+
+    def get_version(self):
+        return None
+
+    def nice_version_name(self):
+        return '<tarball @ {}>'.format(self.tarball)
+
+    def get_tarfile(self):
+        def get_tar_size(TarFile: tarfile.TarFile):
+            return sum(x.size for x in TarFile.getmembers())
+
+        def file_sha1(fp: str):
+            with open(named_temp_file.name, 'rb') as fp:
+                return hashlib.sha1(fp.read()).hexdigest()
+
+        with NamedTemporaryFile(dir=get_downloads_path()) as named_temp_file:
+            # NamedTemporaryFile on top of get_downloads_path
+            # can mean NamedTemporaryFile in TemporaryFile, but works fine
+            logger.debug(f"Using NamedTemporaryFile {named_temp_file.name}")
```

Comment:
We've updated how to we log to use the Events modules.  You'll need to create a new log event for this, and everywhere else you log, and use `fire_event` to kick off writing the log.  We have a [`README`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/events/README.md) with some good info on our logging system.